rfcs/001-mid-level-file-api.md

```diff
@@ -0,0 +1,190 @@
+#[derive(Debug, Clone)]
+pub struct File { ... }
+
+impl File {
+  fn new<T>(
+        name: String,
+        contents: T,
+  ) -> File
+    where
+        T: std::convert::Into<BlobContents>,
+    { ... }
+
+  fn new_with_options<T>(
+        name: String,
+        contents: T,
+        mime_type: Option<String>,
+        last_modified_date: Option<u64>,
+  ) -> File
+    where
+        T: std::convert::Into<BlobContents>,
+    { ... }
+
+  fn name(&self) -> String { ... }
+
+  fn last_modified_since_epoch(&self) -> Duration { ... }
+
+  fn as_blob(&self) -> Blob { ... }
+}
+
+impl BlobLike for File { ... }
+```
+
+`BlobContents` is simply a new-type around `wasm_bindgen::JsValue`s that can be used as the content of `Blob`s and `File`s:
+
+```rust
+#[derive(Debug, Clone)]
+pub struct BlobContents {
+    inner: wasm_bindgen::JsValue,
+}
+```
+
+There are there conversions from types into `BlobContents` only for the types that make sense:
+
+```rust
+impl std::convert::Into<BlobContents> for &str
+impl std::convert::Into<BlobContents> for &[u8]
+impl std::convert::Into<BlobContents> for Blob
+impl std::convert::Into<BlobContents> for js_sys::ArrayBuffer
+```
+
+Lastly there's the `FileReader` which allows reading from BlobLike objects. We'll have two implementations of this, one based on callbacks and the other based on futures.
+
+The callbacks implementation has three categories of callbacks: start, progress and read. Start and progress callbacks are directly related to the `onloadstart` and `onprogress` callbacks on `web_sys::FileReader`. The read variety of callbacks, are a combination of `onload`, `onloadend`, `onloaderror`, and `onabort`. The callback receives a result which is an error if the underlying read was aborted or errored.
+
+The futures implementation likewise exposes success, error, and abort through the fact that futures are `Result`-like. Progress events are exposed as a stream. In the future, we may expose the entire lifecycle of a read through a stream.
+
+```rust
+mod callbacks {
+  #[derive(Debug)]
+  pub struct FileReader { ... }
+
+  impl FileReader {
+    fn new() -> FileReader { ... }
+
+    fn read_as_string<F>(self, blob: &impl BlobLike, callback: F)
+        where F: FnOnce(Result<String, FileReadError>) { ... };
+
+    fn read_as_data_url<F>(self, blob: &impl BlobLike, callback: F)
+        where F: FnOnce(Result<String, FileReadError>) { ... };
+
+    fn read_as_array_buffer<F>(self, blob: &impl BlobLike, callback: F)
+        where F: FnOnce(Result<&web_sys::ArrayBuffer, FileReadError>) { ... };
+
+    fn on_progress<F>(&mut self, callback: F)
+      where F: FnMut(ProgressEvent) + 'static { ... }
+
+    fn on_load_start<F>(&mut self, callback: F)
+      where F: FnOnce(LoadStartEvent) + 'static { ... }
+  }
+}
+
+mod futures {
+  #[derive(Debug)]
+  pub struct FileReader { ... }
+
+  impl FileReader {
+    fn new() -> FileReader { ... }
+
+    fn read_as_string(self, blob: &impl BlobLike) -> ReadAsString { ... }
```

Comment:
@fitzgen These should also probably be called `read_to_*`. Thoughts?

Reply:
Sounds good to me!